storage: use map for watchableStore synced

gyuho opened the pull request:
This is for coreos#3859 switching slice to map for synced watchings.
For a large amount of synced watchings, map implementation performs better.
When putting 1 million watchers on the same key and canceling them one by
one: original implementation takes 9m7.268221091s, while the one with map
takes only 430.531637ms.

gyuho commented:
```
BenchmarkWatchableStoreSyncedCancel:
[1]:
benchmark                                 old ns/op       new ns/op     delta
BenchmarkWatchableStoreSyncedCancel       28481490911     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-2     24699111926     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-4     25444312207     0.07          -100.00%
BenchmarkWatchableStoreSyncedCancel-8     25213736765     0.06          -100.00%

benchmark                                 old allocs     new allocs     delta
BenchmarkWatchableStoreSyncedCancel       5001           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-2     6133           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-4     7427           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-8     8278           0              -100.00%

benchmark                                 old bytes     new bytes     delta
BenchmarkWatchableStoreSyncedCancel       2002296       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-2     1997224       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-4     2126552       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-8     2165512       0             -100.00%


[2]:
benchmark                                 old ns/op       new ns/op     delta
BenchmarkWatchableStoreSyncedCancel       28405861525     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-2     24189639392     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-4     25185842756     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-8     24981059078     0.06          -100.00%

benchmark                                 old allocs     new allocs     delta
BenchmarkWatchableStoreSyncedCancel       5039           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-2     5977           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-4     7258           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-8     8195           0              -100.00%

benchmark                                 old bytes     new bytes     delta
BenchmarkWatchableStoreSyncedCancel       2020232       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-2     1948480       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-4     2092416       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-8     2144696       0             -100.00%


[3]:
benchmark                                 old ns/op       new ns/op     delta
BenchmarkWatchableStoreSyncedCancel       28649823895     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-2     24409337089     0.07          -100.00%
BenchmarkWatchableStoreSyncedCancel-4     25467552651     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-8     25205423967     0.05          -100.00%

benchmark                                 old allocs     new allocs     delta
BenchmarkWatchableStoreSyncedCancel       5020           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-2     6069           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-4     7378           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-8     8310           0              -100.00%

benchmark                                 old bytes     new bytes     delta
BenchmarkWatchableStoreSyncedCancel       2011264       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-2     1977624       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-4     2131168       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-8     2175312       0             -100.00%


[4]:
benchmark                                 old ns/op       new ns/op     delta
BenchmarkWatchableStoreSyncedCancel       28946238621     0.10          -100.00%
BenchmarkWatchableStoreSyncedCancel-2     24380060493     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-4     25614712892     0.05          -100.00%
BenchmarkWatchableStoreSyncedCancel-8     24995822673     0.06          -100.00%

benchmark                                 old allocs     new allocs     delta
BenchmarkWatchableStoreSyncedCancel       4993           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-2     6077           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-4     7357           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-8     8298           0              -100.00%

benchmark                                 old bytes     new bytes     delta
BenchmarkWatchableStoreSyncedCancel       2001784       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-2     1978072       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-4     2114320       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-8     2166792       0             -100.00%


[5]:
benchmark                                 old ns/op       new ns/op     delta
BenchmarkWatchableStoreSyncedCancel       28442078021     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-2     24422697590     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-4     25397499735     0.06          -100.00%
BenchmarkWatchableStoreSyncedCancel-8     25589699437     0.06          -100.00%

benchmark                                 old allocs     new allocs     delta
BenchmarkWatchableStoreSyncedCancel       4978           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-2     6006           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-4     7317           0              -100.00%
BenchmarkWatchableStoreSyncedCancel-8     8427           0              -100.00%

benchmark                                 old bytes     new bytes     delta
BenchmarkWatchableStoreSyncedCancel       1993072       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-2     1958024       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-4     2111760       0             -100.00%
BenchmarkWatchableStoreSyncedCancel-8     2190552       0             -100.00%
```

And the script that I used to measure the 1 million watcher performance:

```
package storage

import (
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"os"
	"path"
	"time"
)

func TEST() {
	tmpDir, err := ioutil.TempDir(os.TempDir(), "etcd_test_storage")
	if err != nil {
		log.Fatal(err)
	}
	tmpPath := path.Join(tmpDir, "database")
	s := newWatchableStore(tmpPath)
	defer func() {
		s.store.Close()
		os.Remove(tmpPath)
	}()

	// Put a key so that we can spawn watchers on that key
	testKey := []byte("foo")
	testValue := []byte("bar")
	s.Put(testKey, testValue)

	w := s.NewWatcher()

	// put 1 million watchers on the same key
	const watcherSize = 1000000

	cancels := make([]CancelFunc, watcherSize)
	for i := 0; i < watcherSize; i++ {
		// 0 for startRev to keep watchers in synced
		cancel := w.Watch(testKey, true, 0)
		cancels[i] = cancel
	}

	// randomly cancel watchers to make it not biased towards
	// data structures with an order, such as slice.
	ix := rand.Perm(watcherSize)

	ss := time.Now()
	for i, idx := range ix {
		if i%10000 == 0 {
			log.Println("Cancelling", i)
		}
		cancels[idx]()
	}
	fmt.Println("took:", time.Since(ss))
}
```

The whole workflow can be found: https://github.com/gyuho/etcd-scripts/tree/master/%233859_synced_map

Thanks,

xiang90 commented on storage/watchable_store.go:
```diff
@@ -185,7 +185,12 @@ func (s *watchableStore) watch(key []byte, prefix bool, startRev int64, ch chan<
 	if startRev == 0 {
-		s.synced[k] = append(s.synced[k], wa)
+		if v, ok := s.synced[k]; ok {
+			v[wa] = struct{}{}
```
let's check if v[wa] exists. It should not exist before or it means we put in the same watch twice. Probably we should just panic if it exists. But please double check.

xiang90 commented on storage/watchable_store.go:
```diff
@@ -203,9 +208,9 @@ func (s *watchableStore) watch(key []byte, prefix bool, startRev int64, ch chan<
-			if w == wa {
-				s.synced[k] = append(s.synced[k][:i], s.synced[k][i+1:]...)
+		if v, ok := s.synced[k]; ok {
+			if _, et := v[wa]; et {
```
et -> ok? ok is more common.

xiang90 commented on storage/watchable_store.go:
```diff
@@ -272,7 +277,13 @@ func (s *watchableStore) syncWatchings() {
 		// switch to tracking future events if needed
 		if nextRev > curRev {
-			s.synced[string(w.key)] = append(s.synced[string(w.key)], w)
+			k := string(w.key)
```
the same code logic repeats twice. Can we make this a nice helper function instead?

xiang90 commented on storage/watchable_store.go:
```diff
@@ -203,9 +208,9 @@ func (s *watchableStore) watch(key []byte, prefix bool, startRev int64, ch chan<
-		for i, w := range s.synced[k] {
-			if w == wa {
-				s.synced[k] = append(s.synced[k][:i], s.synced[k][i+1:]...)
+		if v, ok := s.synced[k]; ok {
```
if there is nothing in s.synced[k], we should also remove s.synced[k] (empty map item)

gyuho replied:
Can you clarify? Like this?

```
			if v, ok := s.synced[k]; ok {
				v[w] = struct{}{}
			} else {
				s.synced[k] = make(map[*watching]struct{})
			}
```

xiang90 replied:
After you cancel the last watching item in synced[k], you also need to remove the map.

xiang90 replied:
It would look like

```go
if len(synced[k]) == 0 {
    delete(synced, k)
}
```

gyuho replied:
Oh got it!

gyuho commented:
Thanks for feedback! I will push updates as you suggest shortly.

xiang90 commented:
@gyuho Thank you!

gyuho commented:
@xiang90 Please review again. I addressed all the issue you mentions.
Performance stays the same. I ran the script on 1 million keys
and map implementation is much faster (9 min vs 400 ms). And
added some test cases.

gyuho commented on storage/watchable_store.go:
```diff
@@ -185,7 +186,7 @@ func (s *watchableStore) watch(key []byte, prefix bool, startRev int64, ch chan<
 	k := string(key)
 	if startRev == 0 {
-		s.synced[k] = append(s.synced[k], wa)
+		_ = safelySynced(s, k, wa)
```
**I need more attention on this line!**

/cc @xiang90

xiang90 commented on storage/watchable_store.go:
```diff
@@ -356,3 +360,25 @@ type watching struct {
+				v[wa] = struct{}{}
+			}
+		}
+		if len(s.synced[k]) == 0 {
```
well... I do not think this should be here. this should be where you REMOVE a watching from the map.

gyuho replied:
Here?

```go
			if _, ok := v[wa]; ok {
				delete(v, wa)
				watchingGauge.Dec()
			}
```

Thanks,

xiang90 commented on storage/watchable_store.go:
```diff
@@ -356,3 +360,25 @@ type watching struct {
+		}
+		if len(s.synced[k]) == 0 {
+			delete(s.synced, k)
+		}
```
let's return here, then we do not need to extra else statement.

xiang90 commented on storage/watchable_store.go:
```diff
@@ -356,3 +360,25 @@ type watching struct {
 }
+
+// safelySynced safely puts watching with key k into watchableStore's synced.
+func safelySynced(s *watchableStore, k string, wa *watching) error {
```
What does safely means? 

Also it is better to make the func a verb. Like PutIntoSynced or AddWatching.
Moreover, why do you need to pass in the watchableStore? I think you only access the map.

gyuho replied:
Got it. I will make it `addWatching`. And change arguments.

xiang90 replied:
actually i think this should be names as Unsafe since you need to hold the lock before doing it, right?

gyuho replied:
Got it. I will make it safe.

xiang90 commented on storage/watchable_store.go:
```diff
@@ -272,7 +285,10 @@ func (s *watchableStore) syncWatchings() {
-			s.synced[string(w.key)] = append(s.synced[string(w.key)], w)
+			k := string(w.key)
+			if err := unsafeAddWatching(&s.synced, k, w); err != nil {
+				plog.Debugf("error unsafeAddWatching (%v) for key", err, k)
```
i think it worth a panic?

xiang90 commented on storage/watchable_store.go:
```diff
@@ -292,25 +308,27 @@ func (s *watchableStore) handle(rev int64, ev storagepb.Event) {
-				s.unsynced[w] = struct{}{}
-				slowWatchingGauge.Inc()
+		k := string(ev.Kv.Key[:i])
+		nwm := make(map[*watching]struct{})
```
i do not think we need to do this. we do not need the new map. we just need to remove the slow one from the map.

gyuho replied:
Got it. Will change.

gyuho commented:
@xiang90 Just changed it to `log.Panicf`. Please take a look again. Thanks,

xiang90 commented on storage/watchable_store_bench_test.go:
```diff
@@ -81,3 +81,39 @@ func BenchmarkWatchableStoreUnsyncedCancel(b *testing.B) {
+	w := s.NewWatcher()
+
+	// put 1 million watchers on the same key
+	const watcherSize = 200000
```
this is 200k, not 1M?

gyuho replied:
I set 200k because it was hanging too long for slice implementation. Will change it for new implementation.

gyuho commented:
@xiang90 Just changed it to 1-million key and here's the result with *map* implementation:

```
test_function_name="XXX"
benchmark_function_name="BenchmarkWatchableStoreSyncedCancel"

go test -run $test_function_name -bench $benchmark_function_name -benchmem -cpu 1,2,4,8
PASS
BenchmarkWatchableStoreSyncedCancel  	2000000000	         0.21 ns/op	       0 B/op	       0 allocs/op
BenchmarkWatchableStoreSyncedCancel-2	2000000000	         0.20 ns/op	       0 B/op	       0 allocs/op
BenchmarkWatchableStoreSyncedCancel-4	2000000000	         0.20 ns/op	       0 B/op	       0 allocs/op
BenchmarkWatchableStoreSyncedCancel-8	2000000000	         0.20 ns/op	       0 B/op	       0 allocs/op
ok  	github.com/coreos/etcd/storage	118.109s

```

xiang90 commented on storage/watchable_store.go:
```diff
@@ -185,7 +186,9 @@ func (s *watchableStore) watch(key []byte, prefix bool, startRev int64, ch chan<
 	if startRev == 0 {
-		s.synced[k] = append(s.synced[k], wa)
+		if err := unsafeAddWatching(&s.synced, k, wa); err != nil {
+			log.Panicf("error unsafeAddWatching (%v) for key %s", err, k)
```
actually can we move this into the unsafeAddWatching func?

gyuho replied:
Sure

xiang90 replied:
after offline discussion, we decided to still return the error for easy testing.

xiang90 commented on storage/watchable_store.go:
```diff
@@ -356,3 +367,24 @@ type watching struct {
+func unsafeAddWatching(synced *map[string]map[*watching]struct{}, k string, wa *watching) error {
+	mp := *synced
+	if v, ok := mp[k]; ok {
+		if wa != nil {
```
hm... if wa is nil, we probably should just panic.

xiang90 replied:
i mean if the code is correct, we should never add a nil watching, right?

gyuho replied:
Agree. I will return error for `nil` watching. Or should I just assume it't not `nil`?
It's useful for some testing when you don't want to create dummy watching. Your call. Thanks,

xiang90 replied:
well... probably the best thing is to return error.

xiang90 replied:
like 

```go
if wa == nil {
   ...
}

...
```

xiang90 replied:
you do not have to check it twice at line 376 and line 385 at least

gyuho replied:
Got it.

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,84 @@
+	}
+}
+
+func TestNewWatcher(t *testing.T) {
```
TestCancelWatch? The first part is already covered by the `TestWatch` test, right?

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,84 @@
+	cancel := w.Watch(testKey, true, 0)
+	cancel()
+	if v, ok := s.synced[string(testKey)]; ok {
+		t.Fatalf("%s shoud have been deleted but %+v", testKey, v)
```
use t.Errorf. Fatalf should be only used when some unexpected thing happen.

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,84 @@
+	w := s.NewWatcher()
+	w.Watch(testKey, true, 0)
+	if v, ok := s.synced[string(testKey)]; !ok {
+		t.Fatalf("%s must have had an entry in synced but %+v", string(testKey), v)
```
use t.Errorf

gyuho commented:
@xiang90 Just updated `nil` handling as you suggested + `t.Errorf` for tests. Thanks,

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,84 @@
+		}
+	}
+	if err := unsafeAddWatching(&s.synced, string(testKey), wa); err == nil {
+		// panics when putting the same watch twice
```
you are not testing panics but errors. you panics outside unsafefunc

gyuho replied:
Will fix comments.

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,91 @@
+		t.Errorf("want true but got %v: %s must have had entry in synced but %+v", ok, testKey, v)
+	} else {
+		if len(v) != 1 {
+			t.Errorf("want 1 but got %d: %s must have ONE entry in its watching map but %v", len(v), testKey, v)
```
`len(v) = %d, want %d` is good enough

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,91 @@
+	}
+
+	if err := unsafeAddWatching(&s.synced, string(testKey), wa); err == nil {
+		t.Error("want <put the same watch twice> but got nil: unsafeAddWatching should have returned error when putting the same watch twice")
```
err = nil, want "put the same watch twice" is good enough

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,91 @@
+	cancel()
+
+	if v, ok := s.synced[string(testKey)]; ok {
+		t.Errorf("want false but got %v: %s shoud have been deleted but %+v", ok, testKey, v)
```
existence = false, want true

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,91 @@
+	w.Watch(testKey, true, 0)
+
+	if v, ok := s.synced[string(testKey)]; !ok {
+		t.Errorf("want true but got %v: %s must have had an entry in synced but %+v", ok, string(testKey), v)
```
existence = false, want true

gyuho commented:
@xiang90 Just updated test outputs. Thanks!

xiang90 commented on storage/watchable_store_test.go:
```diff
@@ -0,0 +1,97 @@
+	} else {
+		if len(v) != 1 {
+			// the key must have ONE entry in its watching map
+			t.Errorf("len(v)=%d, want 1", len(v))
```
can we put a space around `=`? (len(v) = %d)

xiang90 commented:
Test build failed :(

gyuho commented:
Sorry. Will fix shortly!

On Fri, Nov 20, 2015, 18:33 Xiang Li <notifications@github.com> wrote:

> Test build failed :(
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/coreos/etcd/pull/3899#issuecomment-158578170>.
>
-- 
Sincerely,
Gyu-Ho Lee

gyuho commented:
@xiang90 I added white space around `=` and fix the issue found in tests. Will make sure next time. Thanks!

xiang90 commented:
LGTM
